### PR TITLE
Check test naming conventions on all modules

### PR DESCRIPTION
**Pull request by nik9000**
The big win here is catching tests that are incorrectly named and will
be skipped by gradle, providing a false sense of security.

The whole thing takes about 10 seconds on my Macbook Air, not counting
compiling the test classes, which seems worth it. Because this runs as
a gradle task with propery UP-TO-DATE handling it can be skipped if the
tests haven't been changed which should save some time.

I chose to keep this in test:framework rather than a new subproject of
buildSrc because ESIntegTestCase and doesn't inroduce any additional
dependencies.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    boolean skipIntegTestInDisguise = false
+
+    public NamingConventionsTask() {
+        project.afterEvaluate {
+            dependsOn(classpath)
+            description = "Runs NamingConventionsCheck on ${classpath}"
+            executable = new File(project.javaHome, 'bin/java')
+            onlyIf { project.sourceSets.test.output.classesDir.exists() }
+            doFirst({
+                args('-cp', classpath.asPath, 'org.elasticsearch.test.NamingConventionsCheck')
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: Why are you modifying the args during execution? These should all be set during configuration (I thinkt he doFirst can simply be removed?)

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     * Should we skip the integ tests in disguise tests? Defaults to true because only core names its
+     * integ tests correctly.
+     */
+    @Input
+    boolean skipIntegTestInDisguise = false
+
+    public NamingConventionsTask() {
+        project.afterEvaluate {
+            dependsOn(classpath)
+            description = "Runs NamingConventionsCheck on ${classpath}"
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: Many of these lines can be moved out of the afterEvaluate. We should keep afterEvaluate as simple as possible, so it is clear what things actually need to be delayed during configuration (eg depend on a task setting which may not be set yet).

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy [hunk 3]
```diff
@@ -109,4 +110,11 @@ class PrecommitTasks {
         }
         return checkstyleTask
     }
+
+    private static Task configureNamingConventions(Project project) {
+        if (project.sourceSets.findByName("test")) {
+            return project.tasks.create('namingConventionsCheck', NamingConventionsTask)
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: nit: I think we should just call this `namingConventions`, since we don't add "check" to any of our other check tasks.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy (hunk 2 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: Right. This is just a copy and paste error mostly because I don't fully understand afterEvaluate and haven't bothered to figure it out. I'll have a look.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy [hunk 4]
```diff
@@ -46,5 +46,10 @@ public class StandaloneTestPlugin implements Plugin<Project> {
         test.testClassesDir project.sourceSets.test.output.classesDir
         test.mustRunAfter(project.precommit)
         project.check.dependsOn(test)
+
+        project.namingConventionsCheck {
+            // Stand alone tests don't follow the rules!
+            skipIntegTestInDisguise = true
```
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy]: ?? Standalone tests are just unit tests. How do they not follow the rules?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy (hunk 1 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/NamingConventionsTask.groovy]: Ah! This came from cribbing it from JarHellTask. But it needs afterEvaluate because it is filtering the path I think.

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java [hunk 5]
```diff
@@ -31,7 +31,7 @@
 
 import static org.hamcrest.Matchers.equalTo;
 
-public class IngestDocumentMustacheIT extends AbstractMustacheTests {
+public class IngestDocumentMustacheIT extends AbstractMustacheTestCase {
```
Comment (rjernst) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java]: Why is this an IT when it indirectly extends ESTestCase? Why do we have java code running at all in what should be a rest test??

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy (hunk 4 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy]: They also are ESIntegTestCase in some cases. Would you prefer I push this change to each violating stand alone test?

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java (hunk 5 above)
Comment (nik9000) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java]: I don't have an answer for either question. I can certainly track down why this class isn't failing the naming conventions checks though.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy (hunk 4 above)
Comment (rjernst) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy]: Yes, we should make it clear which projects have violators so we can fix them individually, and not allow others to pop up.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy (hunk 3 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: Done.

Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/precommit/PrecommitTasks.groovy]: Or. Done but not yet pushed.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy (hunk 4 above)
Comment (nik9000) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/StandaloneTestPlugin.groovy]: OK - regardless of what you say I'm going to remove this and try and fix those names/apply this param closer to the offending test.

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java (hunk 5 above)
Comment (nik9000) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java]: I think I can tighten up the checks a bit and it'd catch some more weirdness. I'll push something for that in a bit.

**Comment by nik9000** [comment]
OK! I've worked some to clean up the gradle side. It still has some bits that confuse me - namely why I need afterEvaluate AND doFirst to setup the arguments. I know I need one of the two so the classpath is ready, but I haven't a clue why gradle fails without both.

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java (hunk 5 above)
Comment (nik9000) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestDocumentMustacheIT.java]: On second thought lets do that in another PR. Its a nice win just to run the checks for now.

**Comment by nik9000** [comment]
@rjernst can you have another look?

**Comment by rjernst** [comment]
Looks good

**Comment by nik9000** [comment]
Hurray! Thanks for the review @rjernst ! I've merged it!
